Logout

JimFung opened the pull request:
Completes #453 
But more importantly the padding on the delete button is fixed. 

Logout removes the JWT and email from the user's cookies but doesn't remove the token from the server.

JimFung committed:
delete button has proper padding,
oh and i guess we can log out not too

JimFung committed:
delete button has proper padding,
oh and i guess we can log out not too

JimFung committed:
delete button has proper padding,
oh and i guess we can log out not too

JimFung committed:
delete button has proper padding,
oh and i guess we can log out not too

JimFung committed:
delete button has proper padding,
oh and i guess we can log out not too

JimFung committed:
delete button has proper padding,
oh and i guess we can log out not too

JimFung committed:
delete button has proper padding,
oh and i guess we can log out not too

JimFung committed:
delete button has proper padding,
oh and i guess we can log out not too

JimFung committed:
finishing touches

JimFung committed:
finishing touches

JimFung committed:
finishing touches

JimFung committed:
finishing touches

JimFung committed:
finishing touches

JimFung committed:
finishing touches

JimFung committed:
finishing touches

JimFung committed:
finishing touches

JimFung commented on dispatch/static/manager/src/js/index.js:
```diff
@@ -53,6 +53,8 @@ render((
           <Route path='/login' component={Pages.Login} />
         </Route>
 
+        <Route path='/logout' component={Pages.Logout} />
```
I think i put this in the right place, if it should be another level deeper or w.e lmk but i think this looks right.

JimFung commented:
TODO: remove token from server

JimFung committed:
need to reserach how django routes work so that i can overload /auth/token

JimFung committed:
need to reserach how django routes work so that i can overload /auth/token

JimFung committed:
need to reserach how django routes work so that i can overload /auth/token

JimFung committed:
need to reserach how django routes work so that i can overload /auth/token

JimFung committed:
need to reserach how django routes work so that i can overload /auth/token

JimFung committed:
need to reserach how django routes work so that i can overload /auth/token

JimFung committed:
need to reserach how django routes work so that i can overload /auth/token

JimFung committed:
login is working again

JimFung committed:
login is working again

JimFung committed:
login is working again

JimFung committed:
login is working again

JimFung committed:
login is working again

JimFung committed:
login is working again

JimFung commented:
Current plan: 
I want the /auth/token endpoint to manage both login and logout. To achieve this I'll be refactoring the login/logout code so that it'll be in its own viewset. If there is a better way to do this, Give me a shout but this is the direction I'll be going forward with right now.

JimFung committed:
deletes the token on the server now as well

JimFung committed:
deletes the token on the server now as well

JimFung committed:
deletes the token on the server now as well

JimFung committed:
deletes the token on the server now as well

JimFung committed:
deletes the token on the server now as well

JimFung committed:
deletes the token on the server now as well

JimFung committed:
deletes the token on the server now as well

JimFung committed:
deletes the token on the server now as well

psiemens commented on dispatch/apps/api/urls.py:
```diff
@@ -34,6 +34,11 @@
+authentication_management = views.AuthenticationViewSet.as_view({
+    'post': 'user_authenticate',
+    'delete': 'user_unauthenticate'
+})
```
Dope! This is exactly how the `as_view` method is supposed to be used :metal:

JimFung replied:
🐵 see 🐵 do 🎉

psiemens commented on dispatch/apps/api/urls.py:
```diff
@@ -48,5 +53,5 @@
     url(r'^articles/(?P<pk>[0-9]+)/comments/$', article_comments, name='article-comments'),
     # User authorization
-    url(r'^auth/token', views.user_authenticate, name='user-token'),
+    url(r'^auth/token', authentication_management, name='authentication_management'),
```
Can we change this to:

```python
url(r'^auth/token', authentication_management, name='auth-token'),

psiemens commented on dispatch/apps/api/views.py:
```diff
@@ -638,24 +638,35 @@ def list(self, request):
+
+    def user_unauthenticate(self, request):
+
+        user_token = request.data
```
Upon investigation, I don't think we need to be passing the token in the request payload. 

As long as the token header is set (as it already is), DRF will automatically handle authentication and we don't need to access the token string itself. 

I see what you're trying to do here, but I think we can simplify the entire method like so:

```python

def user_unauthenticate(self, request):

    # Get the token for the current user
    # In theory this should never 404, but doesn't hurt to include for safety
    token = get_object_or_404(Token, user=request.user)

    token.delete()

    return Response({}, status=status.HTTP_200_OK)

psiemens commented on dispatch/static/manager/src/js/api/dispatch.js:
```diff
@@ -92,6 +105,9 @@ var DispatchAPI = {
       return postRequest('auth/token', null, payload)
+    },
+    deleteToken: (token) => {
+      return delRequest('auth/token', null, token, token)
```
If we make the backend change above, we don't need to include the token in the payload:

```python
return delRequest('auth/token', null, null, token)
```

psiemens commented on dispatch/static/manager/src/js/pages/LogoutPage.js:
```diff
@@ -0,0 +1,40 @@
+  }
+
+  render(){
+    return (<p>Hasta la vista, baby</p>)
```
Bon voyage 👋

JimFung replied:
👋

blx commented on dispatch/apps/api/views.py:
```diff
@@ -638,24 +638,35 @@ def list(self, request):
-        return Response({}, status=status.HTTP_400_BAD_REQUEST)
+            return Response(data, status=status.HTTP_202_ACCEPTED)
+        else:
+            return Response({}, status=status.HTTP_400_BAD_REQUEST)
```
Recognising that this code was already here, so feel free to ignore:

- Do you think 403 would be a more semantic and standard response here? 400 generally corresponds to "you've sent malformed or bad data", whereas the case here seems more like the request was understood but the server is unwilling to accept it -- 403.
  - https://tools.ietf.org/html/rfc7231#section-6.5.1

- 400 would be appropriate if `email` or `password` or both were not sent at all. On that note, are there going to be cases where attempting to login with just an email or just a password is ok? If not, it would be better for lines 645-646 to look more like:
    ```py
    try:
        email = request.data.get('email')
        password = request.data.get('password')
    except KeyError:
        return Response({}, status=status.HTTP_400_BAD_REQUEST)

blx commented on dispatch/apps/api/views.py:
```diff
@@ -638,24 +638,35 @@ def list(self, request):
+                Token.objects.get(user_token).delete()
+            except:
+                pass
+            return Response({}, status=status.HTTP_200_OK)
```
[200 is the default status](http://www.django-rest-framework.org/api-guide/responses/#response) if none is provided; I think it's more clear to omit 200s, so that the cases where you specify a different status are more obvious, but hey that's up to you.

JimFung replied:
neat, thought i had to explicitly set the status code. I agree with what you said, gonna remove it.

blx commented on dispatch/static/manager/src/js/actions/UserActions.js:
```diff
@@ -46,3 +46,20 @@ export function authenticateUser(email, password, nextPath = '/') {
+  return {
+    type: types.AUTH_DELETE_TOKEN,
+    payload: DispatchAPI.auth.deleteToken(token)
+      .then( json => json )
```
what does this accomplish? isn't this the identity function?

JimFung replied:
🤔 I'll get back to you on that

JimFung replied:
it is the identity function, what does it accomplish? it sparked a mildly interesting back and forth. 👍 taking it out since it's done its job.

JimFung committed:
rebased

JimFung committed:
rebased

JimFung committed:
rebased

JimFung committed:
rebased

JimFung committed:
Merge branch 'logout' of https://github.com/ubyssey/dispatch into logout

JimFung committed:
rolled back version on dep
react-redux-loading-bar 2.6.6 will flood the console with warnings, rolling back to 2.6.5

JimFung committed:
rolled back version on dep
react-redux-loading-bar 2.6.6 will flood the console with warnings, rolling back to 2.6.5

JimFung committed:
changing name to auth-token

JimFung committed:
changing name to auth-token

JimFung committed:
changing name to auth-token

JimFung committed:
changing name to auth-token

JimFung committed:
simplified user_unauthenticate

JimFung committed:
simplified user_unauthenticate

JimFung committed:
not passing token into api call for logout

JimFung committed:
not passing token into api call for logout

JimFung committed:
not passing token into api call for logout

JimFung committed:
removed useless .then

JimFung committed:
Merge branch 'logout' of https://github.com/ubyssey/dispatch into logout

JimFung committed:
specing out the test cases that i need to write

JimFung committed:
specing out the test cases that i need to write

JimFung committed:
filled in some of the testing code

JimFung committed:
filled in some of the testing code

JimFung committed:
one more test left

JimFung committed:
one more test left

JimFung committed:
done tests

JimFung committed:
done tests

JimFung committed:
Merge branch 'logout' of https://github.com/ubyssey/dispatch into logout

JimFung committed:
missed something in the rebase and fixed linting

JimFung committed:
removing old code

psiemens committed:
Prevent token from being accidentally cached

psiemens commented:
I fixed the issue of not being able to log back in immediately after a logout. It turns out that because we weren't creating a copy of the `DEFAULT_HEADERS` global object in `api/dispatch.js`, after the token had been set once it was being persisted until a page refresh. I fixed this by merging the `DEFAULT_HEADERS` object with the user-defined headers to create a fresh object every time.

However, I noticed another bug that I think is worth fixing. If a logout fails, the user is still directed to the login page and the logout page is never shown. After a bit of thought, I came to the conclusion that it doesn't really make sense to have `LogoutPage` at all. Instead, I think that `unauthenticateUser` should be passed as an action to `MainContainer`, which will then pass it onto `Header`. If a logout is successful, the user should be directed to the login page, and if not, an error message should be displayed. 

For an example of how to accomplish this with just actions, check out this code: https://github.com/ubyssey/dispatch/blob/logout/dispatch/static/manager/src/js/actions/UserActions.js#L35-L48

psiemens committed:
Merge branch 'develop' into logout

# Conflicts:
#	dispatch/static/manager/src/js/components/ItemList/ItemListHeader.js
